xnmt/model_base.py

```diff
@@ -1,12 +1,62 @@
-from typing import Sequence, Union
+from typing import Optional, Sequence, Union
 
-from xnmt import batcher, events, input, input_reader, loss, output, training_task
+from xnmt import batcher, events, input_reader, loss, output, training_task
 import xnmt.loss_calculator
 import xnmt.input
+from xnmt.persistence import Serializable, serializable_init
 
 class TrainableModel(object):
   """
-  A template class for a basic trainable model, implementing a loss function based on inputs and outputs.
+  A template class for a basic trainable model, implementing a loss function.
+  """
+
+  def calc_loss(self, *args, **kwargs) -> loss.FactoredLossExpr:
+    """Calculate loss based on input-output pairs.
+
+    Losses are accumulated only across unmasked timesteps in each batch element.
+
+    Arguments are to be defined by subclasses
+
+    Returns:
+      A (possibly batched) expression representing the loss.
+    """
+
+  def get_primary_loss(self) -> str:
+    """
+    Returns:
+      Identifier for primary loss.
+    """
+    raise NotImplementedError("Pick a key for primary loss that is used for dev_loss calculation")
+
+class UnsupervisedModel(TrainableModel):
+  """
+  A template class for an unsupervised trainable model, implementing a loss function based on inputs only.
```

Comment:
I think "Unsupervised" is a bit of a misnomer here. I'd prefer something like "Monolingual" (although I'm not completely satisfied with this either). Also, "based on inputs only" seems a little strange. Maybe "based on a single input" or something?